Allow increments to be added for ocean and ice cycling.

guillaumevernieres opened the pull request:
# Description
Motivation: WCDA cycling. Long description for very few updates.

### Summary of changes:
- Removed forgotten commented out call to the B-mat `j-job` in ```jobs/rocoto/ocnanalbmat.sh``` ... oops
- `MOM6` IAU is now triggered with the `DO_JEDIOCNVAR` switch. Direct insertion can be achieved with the same IAU module.
- Link to the `JEDI/SOCA` increment in `ush/forecast_postdet.sh`
- Fixed a dependency bug in `workflow/rocoto/workflow_tasks.py`
- MOM6 Increment is required in `ICSDIR` for the first 1/2 cycle, changes reflected in  ```workflow/setup_expt.py``` 

### Dependencies
Nothing major, just 1 obs `yaml` updated to the new `ioda` format so we have obs to cycle with.
[GDASApp/pull/349](https://github.com/NOAA-EMC/GDASApp/pull/349)


- Fixes [#346](https://github.com/NOAA-EMC/GDASApp/issues/346)
- Fixes [#350](https://github.com/NOAA-EMC/GDASApp/issues/350)


### Type of change

- [x] New feature (non-breaking change which adds functionality)

### How Has This Been Tested
It only works on `Hera` and with the develop branch of the `GDASApp` repository.
It was tested over 1.5 cycle. I ran out of obs for the GSI and our `ioda` concatenation tool is broken since the merge of the "new naming convention", so we can't do `00z` cycles.

### To reproduce
Edit `config.base`:
```bash
export DOIAU="NO"
export DO_JEDIOCNVAR="YES"
export l4densvar=".false."
export lwrite4danl=".false."
export DO_METP="NO"
```

Resource files needed for a warm start, stored in `ICSDIR`: 
```/scratch2/NCEPDEV/ocean/Guillaume.Vernieres/runs/wcda-tutorial/ICSDIR```, same as @aerorahul 's but with the addition of a MOM6 incr (all zeros)

Create a yaml list of obs `obs_list.yaml`:
```yaml
observers:
- <path to gw>/sorc/gdas.cd/parm/soca/obs/config/adt_j3_egm2008.yaml
```

Configure the ocean/ice DA in `ocnanal.yaml`, the default will be broken until we resolve our obs database issues:
```yaml
ocnanal:
  SOCA_INPUT_FIX_DIR: /scratch2/NCEPDEV/ocean/Guillaume.Vernieres/data/static/72x35x25/soca
  CASE_ANL: 'C24'
  COMIN_OBS: '/scratch2/NCEPDEV/ocean/Guillaume.Vernieres/runs/r2d2-v2-v3'   # Temporary obs file in ioda-v3
  SOCA_OBS_LIST:  <path to your>/obs_list.yaml
  SOCA_NINNER: 1
  R2D2_OBS_SRC: 'gdas_marine'
  R2D2_OBS_DUMP: 's2s_v1'
  SABER_BLOCKS_YAML: ''
  NICAS_RESOL: 1
  NICAS_GRID_SIZE: 15000
```
Setup the experiment:
```bash
./setup_expt.py cycled --app S2S --pslot wcda --configdir $CONFIGDIR --idate 2021032312 --edate 2021032318 --resdet 48 --gfs_cyc 0 --comrot $COMROT --expdir $EXPDIR --nens 0 --start 'warm' --icsdir <path to>/ICSDIR/C48O500 --yaml ocnanal.yaml
```

Generate the `xml` files:
```bash
setup_xml.py $EXPDIR/wcda
```
  
**Checklist**

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

guillaumevernieres committed:
cyc with ocn-ice

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -308,6 +308,9 @@ export imp_physics=@IMP_PHYSICS@
 export DO_JEDIOCNVAR="NO"
+if [[ ${DO_JEDIOCNVAR} = "YES" ]]; then
+    export MOM_IAU=true
+fi
```
There is no need for `MOM_IAU` flag as this is not really an "IAU"
```suggestion
```

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -769,6 +769,11 @@ MOM6_postdet() {
+  # Link increment
+  if [[ "${MOM_IAU}" = "true" ]]; then
+      $NLN "${ROTDIR}/${CDUMP}.${PDY}/${cyc}/ocean/${CDUMP}.t${cyc}z.ocninc.nc" "${DATA}/INPUT/inc.nc"
+  fi
```
```suggestion
  if [[ "${DO_JEDIOCNVAR:-NO}" = "YES" ]]; then
      if [[ ! -f "${ROTDIR}/${CDUMP}.${PDY}/${cyc}/ocean/${CDUMP}.t${cyc}z.ocninc.nc" ]]; then
          echo "FATAL ERROR: Ocean increment not found, ABORT!"
          exit 111
      fi
      $NLN "${ROTDIR}/${CDUMP}.${PDY}/${cyc}/ocean/${CDUMP}.t${cyc}z.ocninc.nc" "${DATA}/INPUT/inc.nc"
      export MOM_IAU="true"
  fi
```

aerorahul reviewed (changes requested):
A few clarifications and updates requested in [MOM_input_template_100](https://github.com/NOAA-EMC/global-workflow/blob/develop/parm/mom6/MOM_input_template_100) and [MOM_input_template_500](https://github.com/NOAA-EMC/global-workflow/blob/develop/parm/mom6/MOM_input_template_500)

The feature is good, just a few inconsistencies in the nomenclature.

guillaumevernieres committed:
Update ush/forecast_postdet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
Update parm/config/config.base.emc.dyn

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
addressed review comments

guillaumevernieres commented:
@aerorahul , I committed your suggestions and addressed your comments about the confusing `MOM_IAU*` variable names.

I also retested the 1.5 cycle.

aerorahul commented on parm/mom6/MOM_input_template_025:
```diff
@@ -833,6 +833,30 @@ ENERGYSAVEDAYS = 1.00           !   [days] default = 1.0
+ODA_INCUPD = @[ODA_INCUPD]   ! [Boolean] default = False
+                             ! If true, oda incremental updates will be applied
+                             ! everywhere in the domain.
+ODA_INCUPD_FILE = "inc.nc"   ! The name of the file with the T,S,h increments.
```
```suggestion
ODA_INCUPD_FILE = "mom6_increment.nc"   ! The name of the file with the T,S,h increments.
```

aerorahul commented on parm/mom6/MOM_input_template_050:
```diff
@@ -858,8 +858,32 @@ USE_NEUTRAL_DIFFUSION = True    !   [Boolean] default = False
+ODA_INCUPD = @[ODA_INCUPD]   ! [Boolean] default = False
+                             ! If true, oda incremental updates will be applied
+                             ! everywhere in the domain.
+ODA_INCUPD_FILE = "inc.nc"   ! The name of the file with the T,S,h increments.
```
```suggestion
ODA_INCUPD_FILE = "mom6_increment.nc"   ! The name of the file with the T,S,h increments.
```

aerorahul commented on parm/mom6/MOM_input_template_100:
```diff
@@ -829,6 +829,28 @@ ENERGYSAVEDAYS = 0.25           !   [days] default = 1.0
+ODA_INCUPD = @[ODA_INCUPD]   ! [Boolean] default = False
+                             ! If true, oda incremental updates will be applied
+                             ! everywhere in the domain.
+ODA_INCUPD_FILE = "inc.nc"   ! The name of the file with the T,S,h increments.
```
```suggestion
ODA_INCUPD_FILE = "mom6_increment.nc"   ! The name of the file with the T,S,h increments.
```

aerorahul commented on parm/mom6/MOM_input_template_500:
```diff
@@ -492,7 +492,9 @@ MAXTRUNC = 1000                 !   [truncations save_interval-1] default = 0
+ODA_INCUPD = @[ODA_INCUPD]   ! [Boolean] default = False
                              ! If true, oda incremental updates will be applied
                              ! everywhere in the domain.
 ODA_INCUPD_FILE = "inc.nc"   ! The name of the file with the T,S,h increments.
```
```suggestion
ODA_INCUPD_FILE = "mom6_increment.nc"   ! The name of the file with the T,S,h increments.
```

aerorahul commented on ush/forecast_postdet.sh:
```diff
@@ -769,6 +769,16 @@ MOM6_postdet() {
+          echo "FATAL ERROR: Ocean increment not found, ABORT!"
+          exit 111
+      fi
+      $NLN "${ROTDIR}/${CDUMP}.${PDY}/${cyc}/ocean/${CDUMP}.t${cyc}z.ocninc.nc" "${DATA}/INPUT/inc.nc"
```
```suggestion
      $NLN "${ROTDIR}/${CDUMP}.${PDY}/${cyc}/ocean/${CDUMP}.t${cyc}z.ocninc.nc" "${DATA}/INPUT/mom6_increment.nc"
```

aerorahul reviewed (approved):
Thanks @guillaumevernieres for incorporating the suggestions.
I have one more suggestion.  The increment file is much better described as `mom6_increment.nc`.  I borrowed the name from the 1 degree `MOM_input_template_100` .  It is in keeping with the atmosphere increment files which are called `fv3_increment.nc` 

The rest of the changes look good.

JessicaMeixner-NOAA commented on parm/mom6/MOM_input_template_100:
```diff
@@ -829,6 +829,28 @@ ENERGYSAVEDAYS = 0.25           !   [days] default = 1.0
                                 ! other globally summed diagnostics.
 
 ! === module ocean_model_init ===
+ODA_INCUPD = @[ODA_INCUPD]   ! [Boolean] default = False
```
So usually the MOM_input_template_* is a copy of what is in ufs-weather-model.  Do these updates need a corresponding test in ufs-weather-model?

guillaumevernieres replied:
Thanks for jumping in @JessicaMeixner-NOAA , I should have thought about adding you when I started modifying the mom namelist! Sorry.
Is there a MOM6 IAU test in the ufs-weather model?

JessicaMeixner-NOAA replied:
I think there is one here: https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/tests/datm_cdeps_iau_gefs but it looks like the settings are different

guillaumevernieres replied:
It seems close enough to me @JessicaMeixner-NOAA but up to you if you think it needs an update to the RT.
@aerorahul was trying to bring the naming nomenclature closer to what the atmosphere DA is doing, which does make sense.

JessicaMeixner-NOAA replied:
That sounds like a great idea to me too.  I think it's important that we are testing in ufs-weather-model what we're actually going to be running and having consistency makes updating the model in the future easier.  I believe there was talk at one point to actually point to the MOM_template files in ufs-eather-model instead of having a second copy of these, but I'm not sure if that's still going to be a future development for g-w or not.  Either way, I just wanted to flag this as something to be aware of, but I don't think it should hold up this PR.

aerorahul replied:
In principle, we should use the templates from the ufs-weather-model for these.
If the settings are different or not templated, then they need to be templated to allow for cycling/workflow v/s model/RT settings.  It might need an update to the RT.

if we use the templates for the coupled model from the ufs-weather-model, is there someone (e.g. from the coupled modeling team) to ensure that when they are updated (new variables added/ old variables removed) in the RT, this information gets relayed to the DA and the workflow teams and/or updated in the workflow configurations?

JessicaMeixner-NOAA replied:
@aerorahul  Whenever I update the version of ufs-weather-model one of the checks is to make sure that the inputs are kept in sync.  With the HR1 PR this is true for ice and ocean except for I haven't added the CICE B-grid, as it's still mostly testing mode.     I think it's a fair statement to say that I'm frequently the person updating the model in the global-workflow and checking namelists is a part of that process.

aerorahul replied:
Thanks @JessicaMeixner-NOAA 
We can let this PR pass and I can open a follow up PR removing these files from global-workflow and using the ones from ufs-weather-model.  
Would you be able to provide a  branch of the ufs-weather-model that allows for this flexibility? 
The ufs-weather-model RT parm/ files may need an update.

JessicaMeixner-NOAA replied:
We'd need the following: 

1. A PR to ufs-weather-model updating  the template files to use these new settings 
2. A PR to global-workflow then updating the model version 
3. A PR having the global-workflow update to then use the ufs-weather-model parm files (I actually thought this was on the workflow teams radar to do?) 

I can do all of the above, but it will be a while before I personally have time to do this.   Again, there's no reason this PR should not proceed.

aerorahul replied:
Yes. 3. is on our radar. 
We need atleast 1. for 3.  
There is a lot more work in the ufs-weather-model than in the global-workflow to achieve this.
I can create issues in the global-workflow and ufs-weather-model and assign us respectively if that is ok?

JessicaMeixner-NOAA replied:
That sounds great.  Just know this isn't anywhere near the top of my priority list at the moment.

aerorahul replied:
@JessicaMeixner-NOAA 
https://github.com/NOAA-EMC/global-workflow/issues/1312
https://github.com/ufs-community/ufs-weather-model/issues/1613
have been created

guillaumevernieres committed:
Update parm/mom6/MOM_input_template_025

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
Update parm/mom6/MOM_input_template_050

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
Update parm/mom6/MOM_input_template_100

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
Update parm/mom6/MOM_input_template_500

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

guillaumevernieres committed:
Update ush/forecast_postdet.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

CoryMartin-NOAA commented on parm/mom6/MOM_input_template_025:
```diff
@@ -868,8 +892,8 @@ LIQUID_RUNOFF_FROM_DATA = @[MOM6_RIVER_RUNOFF]  !   [Boolean] default = False
 DO_SPPT   = @[DO_OCN_SPPT]      ! [Boolean] default = False
-                                ! If true perturb the diabatic tendencies in MOM_diabadic_driver 
-PERT_EPBL = @[PERT_EPBL]        ! [Boolean] default = False 
+                                ! If true perturb the diabatic tendencies in MOM_diabadic_driver
```
`MOM_diabatic_driver` or is it dia*bad*ic?

CoryMartin-NOAA commented on parm/mom6/MOM_input_template_050:
```diff
@@ -913,8 +937,8 @@ LIQUID_RUNOFF_FROM_DATA = @[MOM6_RIVER_RUNOFF]  !   [Boolean] default = False
 DO_SPPT   = @[DO_OCN_SPPT]      ! [Boolean] default = False
-                                ! If true perturb the diabatic tendencies in MOM_diabadic_driver 
-PERT_EPBL = @[PERT_EPBL]        ! [Boolean] default = False 
+                                ! If true perturb the diabatic tendencies in MOM_diabadic_driver
```
again

guillaumevernieres replied:
Maybe somebody had a cold when writing that namelist @CoryMartin-NOAA

CoryMartin-NOAA commented on ush/load_ufsda_modules.sh:
```diff
@@ -1,7 +1,8 @@
 
 ###############################################################
-if [[ "${DEBUG_WORKFLOW:-NO}" == "NO" ]]; then
+export DEBUG_WORKFLOW="${DEBUG_WORKFLOW:-NO}"
```
is this a lint thing?

guillaumevernieres replied:
@CoryMartin-NOAA , nope, that was an issue before you added a default to the other places where `DEBUG_WORKFLOW` is used. I'll remove that line.

CoryMartin-NOAA commented on workflow/setup_expt.py:
```diff
@@ -120,6 +120,14 @@ def link_files_from_src_to_dst(src_dir, dst_dir):
         makedirs_if_missing(dst_dir)
         link_files_from_src_to_dst(src_dir, dst_dir)
 
+        # First 1/2 cycle needs a MOM6 increment
```
why? is it not  a `cold start`?

guillaumevernieres replied:
No, it's a `warm` start.

guillaumevernieres committed:
typos + remove export

CoryMartin-NOAA commented on ush/load_ufsda_modules.sh:
```diff
@@ -1,7 +1,7 @@
 
 ###############################################################
-if [[ "${DEBUG_WORKFLOW:-NO}" == "NO" ]]; then
+if [[ "${DEBUG_WORKFLOW}" == "NO" ]]; then
```
I think this was a mistake to merge this change in...  did anyone test? @jiaruidong2017 seems to be having an issue here
